File: backend/src/merlins_collection/services/triage.py
"""Triage reasons — the one place that decides why an item needs attention.

Triage IS the ``needs_review`` queue (docs/plans/rfc-0008/t11-triage-tab.md). No
second flag exists. What this module adds is the *derived* reasons, which are
computed rather than stored, and the single predicate set shared by the two
consumers that must never disagree:

* ``GET /admin/inventory/search?triage=true`` — the list itself;
* ``GET /admin/triage/counts`` — the sidebar badge.

A badge that counts differently from the list it links to is worse than no badge:
the admin clicks through, finds fewer cards than promised, and stops trusting it.

**Stored vs derived.** ``flagged`` is stored state an admin clears explicitly.
The other two are pure functions of fields the repair tools already write, so
they are *self-healing* — link the card or type the name and the row leaves the
list on its own, with nothing to clean up. That is why they stay computed instead
of being baked into ``needs_review`` at write time.
"""

from __future__ import annotations

from collections.abc import Callable, Iterable

from merlins_collection.models.inventory import (
    MACHINE_REVIEW_REASONS,
    InventoryItem,
    ItemStatus,
    Language,
)


def is_flagged(item: InventoryItem) -> bool:
    """Stored: someone (a human or an importer) put this here on purpose."""
    return item.needs_review


def is_missing_card_id(item: InventoryItem) -> bool:
    """Derived: a catalog-linkable item that never matched a catalog row.

    ``getattr``, not ``item.card_id``: only the raw and graded kinds carry the
    field at all. Sealed product and bulk lots have no catalog link BY DESIGN,
    so a plain attribute access would both raise ``AttributeError`` and, if
    defaulted to None, drag every sealed box into the queue permanently.

    ``no_catalog_match`` is an admin's explicit answer that TCGdex does not carry
    this card. It is STORED precisely because this function is derived: without
    it a human could confirm "there is nothing to match" a hundred times and the
    row would return to the queue on the very next read (RFC 0011 §C).

    Note this is a SUPPRESSION inside an existing predicate, not a fourth
    ``TRIAGE_REASONS`` entry. A new reason would keep the card in Triage, which
    is the opposite of what parking is for. A parked item that is ALSO flagged or
    unnamed keeps those reasons and stays — those are real errors, and "the
    catalog does not have it" is not.
    """
    if not hasattr(item, "card_id"):
        return False
    if item.no_catalog_match:
        return False
    return item.card_id is None


def is_missing_english_name(item: InventoryItem) -> bool:
    """Derived: a Japanese item showing a name a customer cannot read.

    A JP item resolves to a JP catalog row by design (``Language`` docstring), so
    the catalog name is Japanese script and the only fix is an admin-authored
    ``display_name_override`` (T10). Not gated on ``card_id``: an unmatched JP
    item still renders its Japanese ``display_name``.
    """
    return item.language is not Language.EN and item.display_name_override is None


# Ordered so the counts payload and the UI's filter dropdown agree without
# either restating the list. Stored reason first, derived after.
TRIAGE_REASONS: dict[str, Callable[[InventoryItem], bool]] = {
    "flagged": is_flagged,
    "missing_card_id": is_missing_card_id,
    "missing_english_name": is_missing_english_name,
}


def reasons_for(item: InventoryItem) -> list[str]:
    """Every reason this item is in the queue, in ``TRIAGE_REASONS`` order.

    ``needs_triage(item)`` is exactly ``bool(reasons_for(item))``, and the two are
    expressed in terms of each other below so they cannot drift: **a row in the
    list with no reason is the defect this function exists to make impossible**,
    and it is the owner's own report (a WHY column reading empty).

    This is also what the API emits, rather than the frontend recomputing the
    same three rules in TypeScript. That mirror was faithful — it was probed, not
    assumed — which is precisely why the drift would have been silent.
    """
    return [name for name, matches in TRIAGE_REASONS.items() if matches(item)]


def needs_triage(item: InventoryItem) -> bool:
    """True if ANY reason applies — the union the Triage list is built from.

    Every other filter on the admin search is AND-combined, which cannot express
    "flagged OR unlinked OR unnamed". This is the one OR, and it deliberately
    yields each item ONCE however many reasons it qualifies under: parallel
    per-reason queues would show the same card repeatedly and get it "fixed"
    twice.
    """
    return bool(reasons_for(item))


def count_by_reason(items: Iterable[InventoryItem]) -> dict[str, int]:
    """Per-reason hit counts. Reasons overlap, so these do NOT sum to the total."""
    items = list(items)
    return {name: sum(1 for i in items if matches(i)) for name, matches in TRIAGE_REASONS.items()}


# ---------------------------------------------------------------------------
# Scope — which items are even eligible for the queue
# ---------------------------------------------------------------------------

#: An item in one of these states will never be corrected, because the card is
#: gone: it has been sold, lost, or handed back to its consignor. Its data
#: quality is history, not work, and leaving it in a queue whose stated goal is
#: to reach zero means the queue can never get there.
TERMINAL_STATUSES = frozenset({
    ItemStatus.SOLD,
    ItemStatus.LOST,
    ItemStatus.RETURNED_TO_CONSIGNOR,
})


def in_triage_scope(item: InventoryItem, *, include_terminal: bool = False) -> bool:
    """Whether this item is in scope for the queue at the requested breadth.

    Called by BOTH the list and ``GET /admin/triage/counts``, for the same reason
    the predicates above are: a badge that counts a different set from the page
    it links to is worse than no badge. Scoping one of the two and forgetting the
    other is the easiest possible way to reintroduce that.
    """
    return include_terminal or item.status not in TERMINAL_STATUSES


# ---------------------------------------------------------------------------
# Bulk clear — the narrow, safe subset
# ---------------------------------------------------------------------------

#: ``blank_condition`` is EXCLUDED, and this is a money rule, not tidiness. The
#: importer stored ``Condition.NM`` — the most expensive tier — for every card
#: whose condition the sheet left blank, and every customer-facing price scales
#: down from it (LP x0.82, MP x0.58). So an LP card in that state is listed at
#: 1.22x what it is worth and an MP card at 1.72x, in the business's favour.
#: Clearing those flags in bulk silently ratifies that price on every card
#: nobody has checked. Only someone holding the card can fix it, which is why
#: the Triage page grows a condition control instead.
BULK_CLEARABLE_REASONS = frozenset(MACHINE_REVIEW_REASONS - {"blank_condition"})


def is_bulk_clearable(item: InventoryItem) -> bool:
    """Whether a bulk clear may drop this item's ``needs_review`` flag.

    Two conditions, both narrowing:

    1. ``flagged`` is its ONLY reason — an item that is also unlinked or unnamed
       keeps that problem, so clearing the flag would not remove it from the list
       anyway and would only destroy information;
    2. the stored reason is one automation writes. A human's free-text note is
       the one thing in this queue a person deliberately recorded, and a bulk
       button that eats it is strictly worse than no bulk button. Absence of a
       reason is NOT a machine reason: a bare flag carries no evidence of who set
       it, so it is left alone too.
    """
    return reasons_for(item) == ["flagged"] and item.review_reason in BULK_CLEARABLE_REASONS

File: frontend/lib/triage.ts
/**
 * Triage — the shared vocabulary for "this card might be wrong".
 * docs/plans/rfc-0008/t11-triage-tab.md
 *
 * Mirrors `backend/src/merlins_collection/services/triage.py`. The reason keys
 * and the union rule are declared there and restated here only because the two
 * runtimes cannot share a module — if they drift, the list and the sidebar
 * badge disagree about what needs fixing.
 */

/** The wire shape of an inventory row as the admin surfaces receive it. */
export interface TriageItem {
  item_id: string
  kind?: string
  card_id?: string | null
  language?: string | null
  display_name?: string | null
  product_name?: string | null
  description?: string | null
  display_name_override?: string | null
  needs_review?: boolean
  review_reason?: string | null
  condition?: string | null
  condition_modifier?: string | null
  /**
   * Why the SERVER put this row in the queue — the authority, computed by
   * `services/triage.reasons_for`. Optional only because the ordinary admin
   * search does not carry it; `?triage=true` always does.
   */
  triage_reasons?: TriageReason[]
  /**
   * Whether a bulk clear would drop this row's flag. Sent by the server so the
   * confirm dialog can name an exact count without the client re-deriving
   * `MACHINE_REVIEW_REASONS` membership — two copies of that rule is how the
   * button and the endpoint come to disagree about what is being destroyed.
   */
  bulk_clearable?: boolean
  lineage_id?: string | null
  predecessor_item_id?: string | null
  card?: { card_id?: string; name?: string; set_name?: string; number?: string } | null
  // The admin search returns EVERY field on the item; only the ones triage
  // reasons up there are named. The index signature keeps this assignable to
  // `Record<string, unknown>`, which DataTable and CardDetailModal both require
  // (same pattern as `PrepQueueItem` in the Prep Queue page).
  [key: string]: unknown
}

export type TriageReason = 'flagged' | 'missing_card_id' | 'missing_english_name'

/** Human labels for the derived reasons. `flagged` renders its stored note instead. */
export const REASON_LABELS: Record<TriageReason, string> = {
  flagged: 'Flagged for review',
  missing_card_id: 'No catalog link',
  missing_english_name: 'Needs English name',
}

export const REASON_FILTER_OPTIONS: { value: '' | TriageReason; label: string }[] = [
  { value: '', label: 'All reasons' },
  { value: 'flagged', label: REASON_LABELS.flagged },
  { value: 'missing_card_id', label: REASON_LABELS.missing_card_id },
  { value: 'missing_english_name', label: REASON_LABELS.missing_english_name },
]

/**
 * Human labels for the machine keys `review_reason` can hold.
 *
 * `review_reason` is ONE column carrying two different things — a key from
 * `MACHINE_REVIEW_REASONS` (models/inventory.py) written by automation, and
 * free text an admin typed. Rendering the column raw made an imported row read
 * `low_match_confidence`, which tells the person holding the card nothing.
 *
 * Keep this in step with `MACHINE_REVIEW_REASONS`: a key with no entry falls
 * through to the free-text branch and renders as a snake_case token again.
 */
export const MACHINE_REASON_LABELS: Record<string, string> = {
  low_match_confidence: "The matcher wasn't sure this is the right card",
  no_catalog_link: 'Imported with no catalog match',
  blank_condition: 'No condition on the imported row',
  manual_entry: 'Entered by hand',
  // Describes a PSA cert flow RFC 0010 §H drops. Old rows may still carry it,
  // so the label stays; nothing writes it any more.
  cert_lookup_failed: 'Cert lookup failed',
}

/**
 * What to show for a stored `review_reason`, or `null` when there is nothing
 * to show.
 *
 * An admin's own note passes through VERBATIM — it is the one thing in this
 * queue a person deliberately recorded, and rewriting it is worse than showing
 * a machine key. Set membership is what separates the two, exactly as it does
 * on the server, where the same distinction decides what a bulk clear may touch.
 */
export function reviewReasonLabel(reason?: string | null): string | null {
  const trimmed = reason?.trim()
  if (!trimmed) return null
  return MACHINE_REASON_LABELS[trimmed] ?? trimmed
}

/** Only raw and graded items link to a catalog card; sealed and bulk never do. */
function isCardLinkable(item: TriageItem): boolean {
  return item.kind === 'raw' || item.kind === 'graded'
}

/**
 * A PREDICTION of what the server would say, for optimistic updates only.
 *
 * NOT the display authority any more (RFC 0010 T3): the chips a row renders
 * come from `item.triage_reasons`, computed by `services/triage.reasons_for`.
 * This mirror survives solely so the page can predict the next state after a
 * repair — before the server has been asked — and drop a fixed row without a
 * refetch round-trip. If it ever drifts from the Python rules, a row briefly
 * shows the wrong chip and then corrects itself; if it were still the
 * authority, it would show the wrong chip forever.
 */
export function reasonsFor(item: TriageItem): TriageReason[] {
  const reasons: TriageReason[] = []
  if (item.needs_review) reasons.push('flagged')
  if (isCardLinkable(item) && !item.card_id) reasons.push('missing_card_id')
  if (item.language && item.language !== 'EN' && !item.display_name_override) {
    reasons.push('missing_english_name')
  }
  return reasons
}

/**
 * What the CUSTOMER sees, resolved through T10's precedence:
 * `display_name_override -> card.name -> display_name -> card_id -> item_id`.
 *
 * The admin has to see this and not the raw stored name, or they are editing
 * blind against the one field that actually outranks theirs.
 */
export function effectiveName(item: TriageItem): string {
  return (
    item.display_name_override?.trim() ||
    item.card?.name ||
    item.display_name ||
    item.product_name ||
    item.description ||
    item.card_id ||
    item.item_id
  )
}

/**
 * The body the note form sends. `null`, never `''` — the model normalizes blank
 * to None anyway, but an empty string on the wire is a value nobody chose.
 */
export function sendToTriageBody(note?: string | null) {
  const trimmed = note?.trim()
  return { needs_review: true, review_reason: trimmed ? trimmed : null }
}

/**
 * The row-level quick action's body. Deliberately does NOT touch
 * `review_reason`: this is the mid-workflow "that one is wrong" gesture with no
 * note attached, and sending `review_reason: null` here would wipe a reason an
 * earlier pass had recorded.
 */
export function quickFlagBody() {
  return { needs_review: true }
}

/** The body that clears an item. The server stamps `reviewed_at` itself. */
export function clearTriageBody() {
  return { needs_review: false, review_reason: null }
}
